app/src/main/java/org/dicio/dicio_android/eval/SkillEvaluator.java

```diff
@@ -353,14 +372,40 @@ private void evaluateMatchingSkill(final String input) {
             final List<String> inputWords = WordExtractor.extractWords(input);
             final List<String> normalizedWordKeys = WordExtractor.normalizeWords(inputWords);
             final Skill skill = skillRanker.getBest(input, inputWords, normalizedWordKeys);
-
+            if(skill==null) {
+                Log.w("SKILLS", "No matching Skill found");
+                return null;
+            }
             skill.processInput(SkillHandler.getSkillContext());
             return skill;
         })
         .subscribeOn(Schedulers.io())
         .observeOn(AndroidSchedulers.mainThread())
         .subscribe(this::generateOutput, this::onError);
     }
+    private void evaluateMatchingSkill(final String[] inputs) {
+        if (evaluationDisposable != null && !evaluationDisposable.isDisposed()) {
+            evaluationDisposable.dispose();
+        }
+
+        evaluationDisposable = Single.fromCallable(() -> {
+            Skill skill=null;
+            for(String input :inputs){
+                final List<String> inputWords = WordExtractor.extractWords(input);
+                final List<String> normalizedWordKeys = WordExtractor.normalizeWords(inputWords);
+                skill = skillRanker.getBest(input, inputWords, normalizedWordKeys);
+                if (skill != null&&!(skill instanceof TextFallback)) {
+                    break;
+                }
+
+            }
+            skill.processInput(SkillHandler.getSkillContext());
```

Comment:
Just put it here, before calling `processInput`, since `skill.getInput()` contains the last input set.
```suggestion
            displayUserInput(skill.getInput());
            skill.processInput(SkillHandler.getSkillContext());
```

Reply:
As far as i can see there's no getInput() Method in the skill class. Could you implement this, so i can implement all your suggestions?

Reply:
After looking more into it, you are right, there is no `skill.getInput()` and it would not make much sense to implement it since the base `Skill` class does not store the input. Sorry for the confusion